File: packages/benchmarks/trust/elizaos_trust_bench/reporter.py
"""Reporter for the agent trust & security benchmark.

Formats benchmark results as a readable table with per-category metrics,
overall scores, difficulty breakdown, and detailed failure analysis.
"""

from __future__ import annotations

from datetime import datetime, timezone

from elizaos_trust_bench.types import (
    BenchmarkResult,
    DetectionResult,
    TrustTestCase,
)


def _pad_right(s: str, width: int) -> str:
    return s[:width] if len(s) >= width else s + " " * (width - len(s))


def _pad_left(s: str, width: int) -> str:
    return s[:width] if len(s) >= width else " " * (width - len(s)) + s


def _pct(n: float) -> str:
    return f"{n * 100:.1f}%"


def format_report(
    result: BenchmarkResult,
    corpus: list[TrustTestCase],
    detections: list[DetectionResult],
) -> str:
    """Format benchmark results as a human-readable report.

    Args:
        result: The scored benchmark result.
        corpus: The test cases that were run.
        detections: The raw detection results.

    Returns:
        Multi-line string with formatted report.
    """
    det_map: dict[str, DetectionResult] = {d.test_id: d for d in detections}
    lines: list[str] = []

    # Header
    lines.append("")
    lines.append("=" * 80)
    lines.append(f"  Agent Trust & Security Benchmark -- {result.handler_name}")
    ts = datetime.fromtimestamp(result.timestamp, tz=timezone.utc).isoformat()
    lines.append(f"  {ts}")
    lines.append("=" * 80)
    lines.append("")

    # Category summary table
    header = (
        _pad_right("Category", 24)
        + _pad_left("TP", 5)
        + _pad_left("FP", 5)
        + _pad_left("FN", 5)
        + _pad_left("TN", 5)
        + _pad_left("Prec", 8)
        + _pad_left("Recall", 8)
        + _pad_left("F1", 8)
    )
    lines.append(header)
    lines.append("-" * 68)

    for cat in result.categories:
        row = (
            _pad_right(cat.category.value, 24)
            + _pad_left(str(cat.true_positives), 5)
            + _pad_left(str(cat.false_positives), 5)
            + _pad_left(str(cat.false_negatives), 5)
            + _pad_left(str(cat.true_negatives), 5)
            + _pad_left(_pct(cat.precision), 8)
            + _pad_left(_pct(cat.recall), 8)
            + _pad_left(_pct(cat.f1), 8)
        )
        lines.append(row)

    lines.append("-" * 68)
    lines.append("")

    # Overall metrics
    lines.append(f"  Overall Macro F1:        {_pct(result.overall_f1)}")
    lines.append(f"  False Positive Rate:     {_pct(result.false_positive_rate)}")
    lines.append(f"  Total Test Cases:        {result.total_tests}")
    lines.append("")

    # Difficulty breakdown
    db = result.difficulty_breakdown
    lines.append("  Difficulty Breakdown:")
    if db.easy_total > 0:
        lines.append(
            f"    Easy:   {db.easy_correct}/{db.easy_total} "
            f"({db.easy_correct / db.easy_total * 100:.1f}%)"
        )
    if db.medium_total > 0:
        lines.append(
            f"    Medium: {db.medium_correct}/{db.medium_total} "
            f"({db.medium_correct / db.medium_total * 100:.1f}%)"
        )
    if db.hard_total > 0:
        lines.append(
            f"    Hard:   {db.hard_correct}/{db.hard_total} "
            f"({db.hard_correct / db.hard_total * 100:.1f}%)"
        )
    lines.append("")

    # Detailed failures
    misses: list[str] = []
    false_positives: list[str] = []

    for tc in corpus:
        det = det_map.get(tc.id)
        if det is None:
            if tc.expected_malicious:
                misses.append(f"  MISS  [{tc.id}] {tc.description} (no result)")
            continue

        if tc.expected_malicious and not det.detected:
            misses.append(
                f"  MISS  [{tc.id}] {tc.description} (conf: {det.confidence:.2f})"
            )
        elif not tc.expected_malicious and det.detected:
            false_positives.append(
                f"  FP    [{tc.id}] {tc.description} (conf: {det.confidence:.2f})"
            )

    if misses or false_positives:
        lines.append("Failures:")
        lines.append("-" * 80)
        lines.extend(misses)
        lines.extend(false_positives)
        lines.append("")
    else:
        lines.append("No failures -- perfect score!")
        lines.append("")

    lines.append("=" * 80)

    return "\n".join(lines)

File: packages/benchmarks/trust/reporter.ts
/**
 * Reporter for Trust Security Benchmark results.
 * Formats results as a readable table.
 */

import type { BenchmarkResult, CategoryScore, DetectionResult, TestCase } from "./types";

function padRight(str: string, len: number): string {
  return str.length >= len ? str.slice(0, len) : str + " ".repeat(len - str.length);
}

function padLeft(str: string, len: number): string {
  return str.length >= len ? str.slice(0, len) : " ".repeat(len - str.length) + str;
}

function pct(n: number): string {
  return `${(n * 100).toFixed(1)}%`;
}

export function formatReport(
  result: BenchmarkResult,
  handlerName: string,
  corpus: TestCase[],
  detections: DetectionResult[],
): string {
  const lines: string[] = [];
  const detMap = new Map(detections.map((d) => [d.testId, d]));

  lines.push("");
  lines.push("=".repeat(72));
  lines.push(`  Trust Security Benchmark -- ${handlerName}`);
  lines.push(`  ${new Date(result.timestamp).toISOString()}`);
  lines.push("=".repeat(72));
  lines.push("");

  // Category summary table
  lines.push(
    padRight("Category", 22) +
      padLeft("TP", 5) +
      padLeft("FP", 5) +
      padLeft("FN", 5) +
      padLeft("TN", 5) +
      padLeft("Prec", 8) +
      padLeft("Recall", 8) +
      padLeft("F1", 8),
  );
  lines.push("-".repeat(66));

  for (const cat of result.categories) {
    lines.push(
      padRight(cat.category, 22) +
        padLeft(String(cat.truePositives), 5) +
        padLeft(String(cat.falsePositives), 5) +
        padLeft(String(cat.falseNegatives), 5) +
        padLeft(String(cat.trueNegatives), 5) +
        padLeft(pct(cat.precision), 8) +
        padLeft(pct(cat.recall), 8) +
        padLeft(pct(cat.f1), 8),
    );
  }

  lines.push("-".repeat(66));
  lines.push("");
  lines.push(`  Overall Macro F1:     ${pct(result.overallF1)}`);
  lines.push(`  False Positive Rate:  ${pct(result.falsePositiveRate)}`);
  lines.push(`  Total Test Cases:     ${result.totalTests}`);
  lines.push("");

  // Detailed failures
  const failures: string[] = [];
  for (const tc of corpus) {
    const det = detMap.get(tc.id);
    if (!det) continue;

    const shouldDetect = tc.expectedMalicious;
    const wasDetected = det.detected;

    if (shouldDetect && !wasDetected) {
      failures.push(
        `  MISS  [${tc.id}] ${tc.description} (conf: ${det.confidence.toFixed(2)})`,
      );
    } else if (!shouldDetect && wasDetected) {
      failures.push(
        `  FP    [${tc.id}] ${tc.description} (conf: ${det.confidence.toFixed(2)})`,
      );
    }
  }

  if (failures.length > 0) {
    lines.push("Failures:");
    lines.push("-".repeat(72));
    lines.push(...failures);
    lines.push("");
  } else {
    lines.push("No failures -- perfect score!");
    lines.push("");
  }

  lines.push("=".repeat(72));

  return lines.join("\n");
}
